lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -873,15 +641,409 @@ def fetch_gce_metadata(metadata_path)
+    def set_project_id
+      @project_id ||= fetch_gce_metadata('project/project-id') if
+        @platform == Platform::GCE
+      @project_id ||= CredentialsInfo.project_id
+    rescue StandardError => e
+      @log.error 'Failed to obtain project id: ', error: e
+    end
+
+    # 1. Return the value if it is explicitly set in the config already.
+    # 2. If not, try to retrieve it by calling metadata servers directly.
+    def set_vm_id
+      @vm_id ||= fetch_gce_metadata('instance/id') if @platform == Platform::GCE
+      @vm_id ||= ec2_metadata['instanceId'] if @platform == Platform::EC2
+    rescue StandardError => e
+      @log.error 'Failed to obtain vm_id: ', error: e
+    end
+
+    # 1. Return the value if it is explicitly set in the config already.
+    # 2. If not, try to retrieve it locally.
+    def set_vm_name
+      @vm_name ||= Socket.gethostname
+    rescue StandardError => e
+      @log.error 'Failed to obtain vm name: ', error: e
+    end
+
+    # 1. Return the value if it is explicitly set in the config already.
+    # 2. If not, try to retrieve it locally.
+    def set_location
+      # Response format: "projects/<number>/zones/<zone>"
+      @zone ||= fetch_gce_metadata('instance/zone').rpartition('/')[2] if
+        @platform == Platform::GCE
+      @zone ||= 'aws:' + ec2_metadata['availabilityZone'] if
+        @platform == Platform::EC2 && ec2_metadata.key?('availabilityZone')
+    rescue StandardError => e
+      @log.error 'Failed to obtain location: ', error: e
+    end
+
+    # Retrieve monitored resource via the legacy way.
+    #
+    # TODO(qingling128): Use this as only a fallback plan after Metadata Agent
+    # support is added.
+    def determine_agent_level_monitored_resource_via_legacy
+      resource = Google::Apis::LoggingV2beta1::MonitoredResource.new(
+        labels: {})
+      resource.type = determine_agent_level_monitored_resource_type
+      resource.labels = determine_agent_level_monitored_resource_labels(
+        resource.type)
+      resource
+    end
+
+    # Determine agent level monitored resource type.
+    def determine_agent_level_monitored_resource_type
+      case @platform
+      when Platform::OTHER
+        # Unknown platform will be defaulted to GCE instance.
+        return COMPUTE_CONSTANTS[:resource_type]
+
+      when Platform::EC2
+        return EC2_CONSTANTS[:resource_type]
+
+      when Platform::GCE
+        # Resource types determined by @subservice_name config.
+        if @subservice_name
+          SUB_SERVICE_LIST.each do |resource_constants|
+            return resource_constants[:resource_type] if
+              @subservice_name == resource_constants[:service]
+          end
+          # Default back to GCE if invalid value is detected.
+          return COMPUTE_CONSTANTS[:resource_type]
+        end
+
+        # Resource types determined by @detect_subservice config.
+        if @detect_subservice
```

Comment:
I renamed `SUBSERVICE_MAP` to `SUBSERVICE_NAME_MAP`. And added a `SUBSERVICE_ATTRIBUTE_MAP` to add these items (because they are kind of different). The refactored code is as below.

```
      # The map between a resource type and expected subservice attributes.
      SUBSERVICE_ATTRIBUTE_MAP = {
        # GAE.
        APPENGINE_CONSTANTS[:resource_type] =>
          %w(gae_backend_name gae_backend_version),
        # GKE.
        CONTAINER_CONSTANTS[:resource_type] => %w(kube-env),
        # Cloud Dataproc.
        DATAPROC_CONSTANTS[:resource_type] =>
          %w(dataproc-cluster-uuid dataproc-cluster-name)
      }
```
```
        if @detect_subservice
          begin
            attributes = fetch_gce_metadata('instance/attributes/').split
            SUBSERVICE_ATTRIBUTE_MAP.map do |resource_type, expected_attributes|
              return resource_type if
                expected_attributes.all? { |attr| attributes.include?(attr) }
            end
          rescue StandardError => e
            @log.error 'Failed to detect subservice: ', error: e
          end
        end
```

PTAL.